src/ripple/overlay/MessageScheduler.h

```diff
@@ -0,0 +1,496 @@
+ * because `std::weak_ptr` is not equality-comparable with anything,
+ * but we do not want to lock the `std::weak_ptr` to get that ID,
+ * so we copy it.
+ */
+struct Channel
+{
+    Peer::id_t peerId;
+    std::weak_ptr<Peer> peer;
+};
+
+inline Channel
+toChannel(std::shared_ptr<Peer> peer)
+{
+    return {peer->id(), peer};
+}
+
+class MessageScheduler
+{
+public:
+    enum class FailureCode {
+        // The request timed out.
+        TIMEOUT,
+        // The peer disconnected.
+        DISCONNECT,
+        // The message scheduler is shutting down.
+        SHUTDOWN,
+    };
+
+    using RequestId = uint32_t;
+
+    class Offer;
+
+    class Sender;
+
+    // A receiver has methods for success and failure callbacks.
+    // We package them together in the same object because they may share data,
+    // and we want to make it easy to control the lifetime of that data.
+    class Receiver;
+
+    // The `request_cookie` field on a `TMLedgerData` message represents
+    // a peer ID, which is a low number starting at 1 and believed to never
+    // exceed 300. By using a non-overlapping set of numbers for our request
+    // IDs, `PeerImp` can dispatch incoming responses `TMLedgerData` messages to
+    // the correct receiver. Similarly for the `seq` field on
+    // `TMGetObjectByHash` messages.
+    static constexpr RequestId MINIMUM_REQUEST_ID = 1 << 24;
+
+private:
+    using Clock = std::chrono::steady_clock;
+    using Timer = boost::asio::basic_waitable_timer<Clock>;
+    using Time = std::chrono::time_point<Clock>;
+
+    // An in-flight request.
+    struct Request
+    {
+        RequestId id;
+        Channel channel;
+        Receiver* receiver;
+        Timer timer;
+        Time sent;
+    };
+
+    boost::asio::io_service& io_service_;
+    beast::Journal journal_;
+
+    // Offers are negotiations between peers and senders.
+    // This mutex must be locked when handling either set.
+    std::mutex offersMutex_;
+    std::vector<Channel> channels_;
+    // TODO: Use a priority queue.
+    std::vector<Sender*> senders_;
+    bool stopped_ = false;
+
+    // Randomize the first ID to avoid collisions after a restart.
+    // C++23: We may be able to switch to std::randint.
+    // https://en.cppreference.com/w/cpp/experimental/randint
+    std::atomic<RequestId> nextId_ =
+        rand_int<RequestId>(MINIMUM_REQUEST_ID, 1 << 24);
+    std::mutex requestsMutex_;
+    // TODO: Might make sense to use a set instead.
+    hash_map<RequestId, std::unique_ptr<Request>> requests_;
+
+public:
+    MessageScheduler(
+        boost::asio::io_service& io_service,
+        beast::Journal const& journal)
+        : io_service_(io_service), journal_(journal)
+    {
+    }
+
+    /**
+     * If there are any waiting senders, offer these channels to them.
+     * Add any remaining open channels to the pool.
+     *
+     * TODO: Maybe this can be private.
+     */
+    void
+    connect(std::vector<Channel> channels);
```

Comment:
Given we move from `channels` should probably take it by rvalue ref.

Reply:
Will be [changing](https://github.com/XRPLF/rippled/pull/4217#discussion_r1171456492) to non-const reference.